File: packages/python/03-langgraph/src/agent.py
"""LangGraph starter - Graph-based stateful agent workflows."""
from typing import Any, TypedDict, Callable
from enum import Enum


class NodeStatus(Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class AgentState(TypedDict):
    """State that flows through the graph."""
    messages: list[dict]
    current_step: str
    context: dict
    final_answer: str | None


class GraphNode:
    """A node in the agent graph."""
    def __init__(self, name: str, func: Callable[[AgentState], AgentState]):
        self.name = name
        self.func = func
        self.status = NodeStatus.PENDING

    def execute(self, state: AgentState) -> AgentState:
        self.status = NodeStatus.RUNNING
        try:
            result = self.func(state)
            self.status = NodeStatus.COMPLETED
            return result
        except Exception:
            self.status = NodeStatus.FAILED
            raise


class ConditionalEdge:
    """Routes between graph nodes based on state."""
    def __init__(self, condition: Callable[[AgentState], str], routes: dict[str, str]):
        self.condition = condition
        self.routes = routes

    def next_node(self, state: AgentState) -> str:
        key = self.condition(state)
        return self.routes.get(key, "end")


class StateGraph:
    """A directed graph for agent state machines."""
    def __init__(self, state_schema: type):
        self.state_schema = state_schema
        self.nodes: dict[str, GraphNode] = {}
        self.edges: dict[str, str | ConditionalEdge] = {}
        self.entry_point: str | None = None

    def add_node(self, name: str, func: Callable) -> None:
        self.nodes[name] = GraphNode(name, func)

    def add_edge(self, from_node: str, to_node: str) -> None:
        self.edges[from_node] = to_node

    def add_conditional_edges(self, from_node: str, condition: Callable, routes: dict) -> None:
        self.edges[from_node] = ConditionalEdge(condition, routes)

    def set_entry_point(self, node_name: str) -> None:
        self.entry_point = node_name

    def compile(self) -> "CompiledGraph":
        if not self.entry_point:
            raise ValueError("Entry point must be set before compiling")
        return CompiledGraph(self)


class CompiledGraph:
    """A compiled, executable graph."""
    def __init__(self, graph: StateGraph):
        self.graph = graph

    def invoke(self, initial_state: AgentState) -> AgentState:
        state = initial_state.copy()
        current = self.graph.entry_point
        visited = set()
        max_steps = 50

        while current and current != "end" and len(visited) < max_steps:
            if current not in self.graph.nodes:
                break
            node = self.graph.nodes[current]
            state = node.execute(state)
            visited.add(current)
            state["current_step"] = current

            edge = self.graph.edges.get(current)
            if edge is None:
                break
            elif isinstance(edge, ConditionalEdge):
                current = edge.next_node(state)
            else:
                current = edge

        return state

File: packages/typescript/16-langgraph-js/src/agent.ts
/**
 * LangGraph.js starter - Graph-based stateful agent workflows in TypeScript
 */

export type NodeFunction<S extends Record<string, unknown>> = (state: S) => Promise<Partial<S>>
export type EdgeCondition<S extends Record<string, unknown>> = (state: S) => string

export interface GraphNode<S extends Record<string, unknown>> {
  name: string
  fn: NodeFunction<S>
}

export interface ConditionalEdge<S extends Record<string, unknown>> {
  from: string
  condition: EdgeCondition<S>
  routes: Record<string, string>
}

export class StateGraph<S extends Record<string, unknown>> {
  private nodes = new Map<string, NodeFunction<S>>()
  private edges = new Map<string, string>()
  private conditionalEdges: ConditionalEdge<S>[] = []
  private entryPoint?: string

  addNode(name: string, fn: NodeFunction<S>): this {
    this.nodes.set(name, fn)
    return this
  }

  addEdge(from: string, to: string): this {
    this.edges.set(from, to)
    return this
  }

  addConditionalEdges(from: string, condition: EdgeCondition<S>, routes: Record<string, string>): this {
    this.conditionalEdges.push({ from, condition, routes })
    return this
  }

  setEntryPoint(name: string): this {
    this.entryPoint = name
    return this
  }

  compile(): CompiledStateGraph<S> {
    if (!this.entryPoint) throw new Error('Entry point not set')
    return new CompiledStateGraph(this.nodes, this.edges, this.conditionalEdges, this.entryPoint)
  }
}

export class CompiledStateGraph<S extends Record<string, unknown>> {
  constructor(
    private nodes: Map<string, NodeFunction<S>>,
    private edges: Map<string, string>,
    private conditionalEdges: ConditionalEdge<S>[],
    private entryPoint: string
  ) {}

  async invoke(initialState: S): Promise<S> {
    let state = { ...initialState }
    let current: string | undefined = this.entryPoint
    const visited = new Set<string>()

    while (current && current !== '__end__' && visited.size < 50) {
      const fn = this.nodes.get(current)
      if (!fn) break
      const updates = await fn(state)
      state = { ...state, ...updates }
      visited.add(current)

      const condEdge = this.conditionalEdges.find(e => e.from === current)
      if (condEdge) {
        const key = condEdge.condition(state)
        current = condEdge.routes[key] ?? '__end__'
      } else {
        current = this.edges.get(current)
      }
    }
    return state
  }

  get nodeCount(): number {
    return this.nodes.size
  }
}

export interface AgentState {
  messages: Array<{ role: string; content: string }>
  context: Record<string, unknown>
  answer?: string
}

export function createReActGraph(llm?: { complete: (prompt: string) => Promise<string> }) {
  const graph = new StateGraph<AgentState>()

  graph.addNode('think', async (state) => {
    if (!llm) return state
    const lastMsg = state.messages[state.messages.length - 1]?.content ?? ''
    const thought = await llm.complete(`Think step by step: ${lastMsg}`)
    return { context: { ...state.context, thought } }
  })

  graph.addNode('answer', async (state) => {
    if (!llm) return { answer: 'No LLM provided' }
    const thought = state.context.thought as string ?? ''
    const ans = await llm.complete(`Based on thinking: ${thought}\nProvide answer:`)
    return { answer: ans }
  })

  graph.addEdge('think', 'answer')
  graph.setEntryPoint('think')
  return graph.compile()
}
